crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -179,17 +196,27 @@ impl Paginator {
             });
         }
 
+        let reset_state_guard =
+            ResetStateGuard { target: PaginatorState::Initial, state: self.state.clone() };
+
         // TODO: do we want to lazy load members?
         let lazy_load_members = true;
 
         let response =
             self.room.event_with_context(event_id, lazy_load_members, num_events).await?;
 
+        // NOTE: it's super important to not have any `await` after this point, since we
+        // don't want the task to be interrupted anymore, or the internal state
+        // may become incorrect.
+
         let has_prev = response.prev_batch_token.is_some();
         let has_next = response.next_batch_token.is_some();
-        *self.prev_batch_token.lock().await = response.prev_batch_token;
-        *self.next_batch_token.lock().await = response.next_batch_token;
+        *self.prev_batch_token.lock().unwrap() = response.prev_batch_token;
+        *self.next_batch_token.lock().unwrap() = response.next_batch_token;
```

Comment:
Do we want these two to perhaps share the lock?

Reply:
Hmm, we're very close to being able to paginate backwards *while* we paginate forwards, by using two locks… I'd like to keep it that way and implement that later on.

Reply:
I realize it's probably premature optimization, but also it's nice for the `paginate()` internal helper to have two locks. Considering they're all fine-grained and shouldn't be contended in general, seems OK to keep them both?

Reply:
Sure, I just wasn't sure why we'd not have two locks considering the above code snippet.